### ISSUE TITLE
[JENKINS-9665] Variables used as node labes are not expanded in parameterized builds

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-06-a2#10943]
<p>Our build process is split into several parameterized jobs. To make sure that these build steps are executed on a specific node, we defined a parameter with the node name the job should be executed on. Unfortunately the variable doesn't get expanded.</p>

<p>Proposed solution:</p>

<p>Values entered as labels should be expanded.</p>

**Attachments:**

- [node.png](https://issues.jenkins.io/secure/attachment/20472/node.png)

---
<details><summary><i>Originally reported by <a title="Maik Richey" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=compi">compi</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-9665" target="_blank">Variables used as node labes are not expanded in parameterized builds</a></i></summary>
<i><ul>
<li><b>reported by</b>: compi
<li><b>status</b>: Closed
<li><b>priority</b>: Critical
<li><b>resolution</b>: Won't Do
<li><b>resolved</b>: 2025-05-19T13:50:12+00:00
<li><b>votes</b>: 13
<li><b>watchers</b>: 18
<li><b>imported</b>: 2025-11-06
</ul></i>
</details>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=compi">compi</a>:</i>
<p>I would like to give some more details. Given are the projects A, B, C, D, E and F.</p>

<p>1. A checks out the workspace, builds it and does an install to the local repository.<br/>
2. B gets then triggered n times, using parts of the workspace of A using the Clone Workspace SCM Plugin and generate large parts of our software. B gets executed in parallel on different nodes and copy what's generated back to the workspace of the node where A got executed on.<br/>
3. When all jobs are finished the joining job C gets triggered. C then triggers D, E and F.<br/>
4. D, E and F are executed sequentially using the workspace of A (including whats generated during the execution of B). To not set the node in each job to a specific one directly it is necessary to define the node to run on as a variable and pass it from one job to another.</p>

<p><b>Problem:</b><br/>
Problem is, that variables used as label for a node do not get expanded to the value given as parameter.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=compi">compi</a>:</i>

<p><b>Solution:</b><br/>
Values entered as labels should be expanded for parameterized builds to keep track of the needs mentioned above.</p>


<p><b>Alternative solution:</b><br/>
If downstream jobs would get executed on the same node the upstream job was executed on when you leave the field "Label Expression" empty (instead of using the Master) then it should work as well. In that case you do not have the chance to define anything as "Label Expression" indeed.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=splosion">splosion</a>:</i>
<p>I would like to second the Alternative Solution.  I have several jobs that would benefit from this change.  It would greatly simplify creating generic steps for multiple jobs.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=compi">compi</a>:</i>
<p>Have a look at the <a href="https://wiki.jenkins-ci.org/display/JENKINS/Parameterized+Trigger+Plugin" class="external-link" target="_blank" rel="nofollow noopener">Parameterized Trigger Plugin</a> and the <a href="https://wiki.jenkins-ci.org/display/JENKINS/NodeLabel+Parameter+Plugin" class="external-link" target="_blank" rel="nofollow noopener">NodeLabel Parameter Plugin</a>. Using both can give what you are looking for. Notice that you need at least Jenkins 1.417 in order to use the NodeLabel Parameter Plugin.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev">oleg_nenashev</a>:</i>
<p>Maik confirmed that "new" plugins resolved his issue</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=danielbeck">danielbeck</a>:</i>
<p>Node Label Parameter Plugin and Parameterized Trigger Plugin resolve this issue.</p>

<p>If you can't get them to work together, please file a new issue against these plugins, preferably after having asked for help on the jenkinsci-users mailing list and having read <a href="https://wiki.jenkins-ci.org/display/JENKINS/How+to+report+an+issue" class="external-link" target="_blank" rel="nofollow noopener">https://wiki.jenkins-ci.org/display/JENKINS/How+to+report+an+issue</a></p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=bizdelnick">bizdelnick</a>:</i>
<p>Hi!<br/>
I reopen this issue because suggested workaround is not acceptable for me. First, NodeLabel Parameter plugin does not allow to enter label expression for cloud. It says "The label expression "trusty" does not match any node" even if it matches one or more clouds. Second, it is very inconvenient to use. Person running the build shouldn't have to remember which labels are available, and the correct expression can be fairly complex. The best solution would be to use a choise parameter and substitute a variable in the label expression.<br/>
For example, I use a label expression like "trusty && my_tools && docker" that is served by a cloud. At some point I need to add as alternative "xenial && my_tools && docker" that is served by another cloud. I want to use a choise parameter "DISTRO" with two possible values "trusty" and "xenial" and set label expression to "${DISTRO} && my_tools && docker".</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=knightsg">knightsg</a>:</i>
<p>We also have this requirement, using variables as content for the node label. The NodeLabel parameter plugin is not a solution as it also doesn't allow the use of variables as the default parameter value, the string entered has to match an existing node name.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rodee">rodee</a>:</i>
<p>+1 from me, in my case, upstream jobs decides where to run the donwstream job, this feature will be very useful.</p>
